libraries/AP_HAL_ChibiOS/hwdef/common/usbcfg_dualcdc.c

```diff
@@ -0,0 +1,559 @@
+  USB_DESC_INTERFACE_ASSOCIATION(                                           \
+    comIfNum,                               /* bFirstInterface.         */  \
+    2,                                      /* bInterfaceCount.         */  \
+    CDC_COMMUNICATION_INTERFACE_CLASS,      /* bFunctionClass.          */  \
+    CDC_ABSTRACT_CONTROL_MODEL,             /* bFunctionSubClass.       */  \
+    1,                                      /* bFunctionProcotol.       */  \
+    0                                       /* iInterface.              */  \
+  ),                                                                        \
+  /* CDC Interface descriptor set */                                        \
+  CDC_IF_DESC_SET(comIfNum, datIfNum, comInEp, datOutEp, datInEp)
+
+/* Configuration Descriptor tree for a CDC.*/
+static const uint8_t vcom_configuration_descriptor_data[] = {
+  /* Configuration Descriptor.*/
+  USB_DESC_CONFIGURATION(
+    USB_DESC_CONFIGURATION_SIZE +
+    (IAD_CDC_IF_DESC_SET_SIZE * 2),         /* wTotalLength.                */
+    USB_NUM_INTERFACES,                     /* bNumInterfaces.              */
+    0x01,                                   /* bConfigurationValue.         */
+    0,                                      /* iConfiguration.              */
+    0xC0,                                   /* bmAttributes (self powered). */
+    50                                      /* bMaxPower (100mA).           */
+  ),
+  IAD_CDC_IF_DESC_SET(
+    USB_CDC_CIF_NUM0,
+    USB_CDC_DIF_NUM0,
+    USB_ENDPOINT_IN(USB_INTERRUPT_REQUEST_EP_A),
+    USB_ENDPOINT_OUT(USB_DATA_AVAILABLE_EP_A),
+    USB_ENDPOINT_IN(USB_DATA_REQUEST_EP_A)
+  ),
+  IAD_CDC_IF_DESC_SET(
+    USB_CDC_CIF_NUM1,
+    USB_CDC_DIF_NUM1,
+    USB_ENDPOINT_IN(USB_INTERRUPT_REQUEST_EP_B),
+    USB_ENDPOINT_OUT(USB_DATA_AVAILABLE_EP_B),
+    USB_ENDPOINT_IN(USB_DATA_REQUEST_EP_B)
+  ),
+};
+
+
+/*
+ * Configuration Descriptor wrapper.
+ */
+static const USBDescriptor vcom_configuration_descriptor = {
+  sizeof vcom_configuration_descriptor_data,
+  vcom_configuration_descriptor_data
+};
+
+/*
+ * U.S. English language identifier.
+ */
+static const uint8_t vcom_string0[] = {
+  USB_DESC_BYTE(4),                     /* bLength.                         */
+  USB_DESC_BYTE(USB_DESCRIPTOR_STRING), /* bDescriptorType.                 */
+  USB_DESC_WORD(0x0409)                 /* wLANGID (U.S. English).          */
+};
+
+/*
+ * Strings wrappers array. The strings are created dynamically to
+ * allow them to be setup with apj_tool
+ */
+static USBDescriptor vcom_strings[] = {
+  {sizeof vcom_string0, vcom_string0},
+  {0, NULL}, // manufacturer
+  {0, NULL}, // product
+  {0, NULL}, // version
+};
+
+#define USB_DESC_MAX_STRLEN 100
+static uint8_t vcom_buffers[3][2+2*USB_DESC_MAX_STRLEN];
+
+/*
+  check if one string contains another
+ */
+static bool string_contains(const char *haystack, const char *needle)
```

Comment:
we should break out the common functions like this so they aren't duplicated. Maybe usb_common.cpp ?

Reply:
Yes, will do.